enhancements/cluster-logging/elasticsearch-v2-api-and-infrastructure.md

```diff
@@ -0,0 +1,310 @@
+---
+title: elasticsearch-v2-api-and-infrastructure
+authors:
+  - "@ewolinetz"
+reviewers:
+  - "@jcantrill"
+  - "@bparees"
+  - "@alanconway"
+approvers:
+  - "@jcantrill"
+  - "@bparees"
+  - "@alanconway"
+creation-date: 2021-07-02
+last-updated: 2021-06-02
+status: implementable
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# elasticsearch-v2-api-and-infrastructure
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Migration plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently when interacting with the Elasticsearch Operator via the Elasticsearch CRD, one needs to have a lot of knowledge on how to best configure
+the Elasticsearch cluster node's roles, memory/cpu, and replication count. In order to continue to improve the user experience of the operator
+and provide a performant cluster that can operate at scale we should simplify the API and move the knowledge of desired infrastructure into the
+operator itself.
+
+## Motivation
+
+Currently the majority of our Elasticsearch bugs are performance related and are usually clusters that have been scaled vertically rather than horizontally.
+Given each node responsibility of Elasticsearch uses its memory differently, having nodes that act as "All-In-One" causes resources to not be used
+in optimal ways and usually requires an over allocation of resources (which makes it hard to justify horizontal scaling, which leads to performance
```

Comment:
It is not clear to me how this motivation maps to the changes proposed.

It "looks" like the ability specify individual roles is being removed in-favor of a small set of generic categories in which all the smarts for how to use Elasticsearch are placed.

> Currently the majority of our Elasticsearch bugs are performance related and are usually clusters that have been scaled vertically rather than horizontally.

How will this interface change address this?  Is it by providing a set of "t-shirt sizes" that map more closely to what the user wants so that they can scale correctly?  If so, then we might want to say that following this sentence.

What is nice about that is the t-shirt sizes in the operator can be adjusted over time as necessary.

What is difficult about that is when the t-shirt doesn't fit for the use case.

> Given each node responsibility of Elasticsearch uses its memory differently, having nodes that act as "All-In-One" causes resources to not be used in optimal ways and usually requires an over allocation of resources (which makes it hard to justify horizontal scaling, which leads to performance related issues).

There is a lot in this sentence.  It seems that what we are saying here is that is too difficult to express what the user needs for their environment with the existing API, and/or the user can easily make mistakes which cause performance problems.  So the motivation is that the new API helps prevent the user from shooting themselves in the foot by reducing the problem to picking the appropriate t-shirt size.

This is a laudable goal.  So then it is all about providing the proper t-shirt sizes.